source/common/crypto/utility_impl.cc

```diff
@@ -71,10 +76,118 @@ const VerificationOutput UtilityImpl::verifySignature(absl::string_view hash, Cr
   return {false, absl::StrCat("Failed to verify digest. Error code: ", ok)};
 }
 
+const SignOutput UtilityImpl::sign(absl::string_view hash, CryptoObject& key,
+                                   const std::vector<uint8_t>& text) {
+  // Sign data using a private key
+  // The key must be imported via importPrivateKey() which supports both DER and PEM formats
+  // Step 1: initialize EVP_MD_CTX
+  bssl::ScopedEVP_MD_CTX ctx;
+
+  // Step 2: initialize EVP_MD
+  const EVP_MD* md = getHashFunction(hash);
+
+  if (md == nullptr) {
+    return {false, {}, absl::StrCat(hash, " is not supported.")};
+  }
+
+  // Step 3: initialize EVP_DigestSign
+  auto pkey_wrapper = Common::Crypto::Access::getTyped<Common::Crypto::PrivateKeyObject>(key);
+  if (pkey_wrapper == nullptr) {
+    return {false, {}, "Failed to initialize digest sign."};
+  }
+  EVP_PKEY* pkey = pkey_wrapper->getEVP_PKEY();
+
+  if (pkey == nullptr) {
+    return {false, {}, "Failed to initialize digest sign."};
+  }
+
+  int ok = EVP_DigestSignInit(ctx.get(), nullptr, md, nullptr, pkey);
+  if (!ok) {
+    return {false, {}, "Failed to initialize digest sign."};
+  }
+
+  // Step 4: get signature length
+  size_t sig_len = 0;
+  ok = EVP_DigestSign(ctx.get(), nullptr, &sig_len, text.data(), text.size());
+  if (!ok) {
+    return {false, {}, "Failed to get signature length."};
+  }
+
+  // Step 5: create signature
+  std::vector<uint8_t> signature(sig_len);
+  ok = EVP_DigestSign(ctx.get(), signature.data(), &sig_len, text.data(), text.size());
```

Comment:
Not a crypto expert, but it's odd to call this twice, and might be expensive to run PKI instructions twice.